lib/streamlit/runtime/caching/cache_utils.py

```diff
@@ -248,7 +248,28 @@ def get_or_create_cached_value():
                         return_value = func(*args, **kwargs)
 
                 messages = cached_func.message_call_stack._most_recent_messages
-                cache.write_result(value_key, return_value, messages)
+                try:
+                    cache.write_result(value_key, return_value, messages)
+                except TypeError as err:
+                    if (
+                        return_value.__module__ == "snowflake.snowpark.session"
+                        and type(return_value).__name__ == "Session"
+                    ):
+
+                        class SnowparkSessionInMemoDecoratorError(TypeError):
+                            def __init__(self):
+                                TypeError.__init__(
+                                    self,
+                                    """
+                                    The function snowflake.snowpark.session.Session is cached with st.memo,
```

Comment:
Let's try to keep wording consistent with this doc: https://www.notion.so/streamlit/Spec-9497a30d456f474eaa2343032ef8cb48#2153548903b14c89b0eede7530d0ca06

Final wording you can discuss with @jrieke !